### PR TITLE
Use Attribute to designate wave-sensitive intrinsics

### PR DESCRIPTION
**Pull request by pow2clk**
This adds an intrinsic attribute to indicate wave-sensitivity that can
be indicated in gen_intrin_main.txt. This and other attributes are
passed along through function representations and lowerings. The
wave-sensitivity needs to be maintained specifically through SROA passes
since it is used by the CleanupDxbreak pass that comes after them.

Specifically this is done to allow extension intrinsics to indicate
wave-sensitivity, but the same mechanism is now used for builtin
intrinsics.

Most intrinsics get a mostly nameless function to represent them between
Codegen and dxilgen. This allows any that have the same prototype to
share the same function. For wave-sensitive intrinsics, they need a
different function or else the attrubute would be similarly shared with
intrinsics matching the prototype. So a minor change is made to their
function names to prevent this.

Adds testing for all these ops and a dummy extension one.

### CODE REVIEW

File: lib/HLSL/DxilPreparePasses.cpp [hunk 1]
```diff
@@ -1142,14 +1142,14 @@ class CleanupDxBreak : public FunctionPass {
...
-        CallInst *CI = dyn_cast<CallInst>(U);
-        if (CI && IsCallWaveSensitive(CI))
-          CollectSensitiveBlocks(LInfo, CI, BreakFunc, SensitiveBBs);
+      if (&IF != &F && IF.getNumUses() && IF.isDeclaration() &&
+          (hlsl::GetHLOpcodeGroupByName(&IF) == HLOpcodeGroup::HLIntrinsic ||
+           hlsl::GetHLOpcodeGroup(&IF) == HLOpcodeGroup::HLExtIntrinsic)) {
+        for (User *U : IF.users()) {
+          CallInst *CI = dyn_cast<CallInst>(U);
+          if (CI && IsHLWaveSensitive(CI->getCalledFunction()))
+            CollectSensitiveBlocks(LInfo, CI, BreakFunc, SensitiveBBs);
```
Comment (pow2clk) [inline lib/HLSL/DxilPreparePasses.cpp:1151]: The addition here is the HLExtInstrinsic check. I inverted the conditionals and removed the continue because I found it easier to follow rather than all the negations.

File: lib/HLSL/HLMatrixLowerPass.cpp [hunk 2]
```diff
@@ -834,24 +834,32 @@ Value *HLMatrixLowerPass::lowerHLIntrinsic(CallInst *Call, IntrinsicOp Opcode) {
...
-      LoweredArgs.emplace_back(getLoweredByValOperand(Arg, Builder));
+      LoweredArg = getLoweredByValOperand(Arg, Builder);
     }
+    LoweredArgs.emplace_back(LoweredArg);
+    ArgTys.emplace_back(LoweredArg->getType());
   }
 
   Type *LoweredRetTy = HLMatrixType::getLoweredType(Call->getType());
-  return callHLFunction(*m_pModule, HLOpcodeGroup::HLIntrinsic, static_cast<unsigned>(Opcode), 
-    LoweredRetTy, LoweredArgs, Builder);
```
Comment (pow2clk) [inline lib/HLSL/HLMatrixLowerPass.cpp:854]: Much of the new code in this area was lifted out of callHLFunction so I could access the call to GetOrCreateHLFunction and provide attributes.

File: lib/HLSL/HLOperations.cpp [hunk 3]
```diff
@@ -289,6 +292,17 @@ void SetHLLowerStrategy(Function *F, StringRef S) {
   F->addFnAttr(HLLowerStrategy, S);
 }
 
+// Set function attribute indicating wave-sensitivity
+void SetHLWaveSensitive(Function *F) {
+  F->addFnAttr(HLWaveSensitive, "y");
```
Comment (pow2clk) [inline lib/HLSL/HLOperations.cpp:297]: The only checks made on this attribute are whether it's present. The "y" doesn't matter, but it's informative.

File: lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp [hunk 4]
```diff
@@ -993,7 +993,7 @@ static unsigned IsPtrUsedByLoweredFn(
 
     } else if (ConstantExpr *CE = dyn_cast<ConstantExpr>(user)) {
       unsigned opcode = CE->getOpcode();
-      if (opcode == Instruction::AddrSpaceCast || Instruction::GetElementPtr)
+      if (opcode == Instruction::AddrSpaceCast || opcode == Instruction::GetElementPtr)
```
Comment (pow2clk) [inline lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp:996]: This is a completely unrelated change that I noticed and added incidentally since the previous conditional was always true, I'm certain it wasn't what was intended.

File: tools/clang/test/HLSLFileCheck/hlsl/intrinsics/wave/reduction/WaveAndBreakLib.hlsl [hunk 5]
```diff
@@ -4,6 +4,8 @@ StructuredBuffer<int> buf[]: register(t2);
 
 // Cannonical example. Expected to keep the block in loop
 // Verify this function loads the global
+// CHECK: define i32
+// CHECK-SAME: WaveInLoop
```
Comment (pow2clk) [inline tools/clang/test/HLSLFileCheck/hlsl/intrinsics/wave/reduction/WaveAndBreakLib.hlsl:8]: Stole these from my other pending dx.break change to make errors clearer by preventing later functions from satisfying CHECK criteria.

File: lib/HLSL/HLOperations.cpp [hunk 6]
```diff
@@ -510,6 +500,9 @@ Function *GetOrCreateHLFunction(Module &M, FunctionType *funcTy,
   }
   else {
     mangledNameStr << GetHLFullName(group, opcode);
+    // Need to add wave sensitivity to name to prevent clashes with non-wave intrinsic
+    if(attribs.hasAttribute(AttributeSet::FunctionIndex, HLWaveSensitive))
+        mangledNameStr << "waveOp";
```
Comment (pow2clk) [inline lib/HLSL/HLOperations.cpp:505]: See WaveProtMatch.hlsl test below for justification for this.

File: include/dxc/HLSL/HLOperations.h [hunk 7]
```diff
@@ -389,6 +390,18 @@ llvm::Function *GetOrCreateHLFunction(llvm::Module &M,
...
+                                      HLOpcodeGroup group, unsigned opcode,
+                                      const llvm::AttributeSet &attribs);
+llvm::Function *GetOrCreateHLFunction(llvm::Module &M,
+                                      llvm::FunctionType *funcTy,
+                                      HLOpcodeGroup group,
+                                      llvm::StringRef *groupName,
+                                      llvm::StringRef *fnName,
+                                      unsigned opcode,
+                                      const llvm::AttributeSet &attribs);
+
```
Comment (pow2clk) [inline include/dxc/HLSL/HLOperations.h:404]: These variants allow providing attribs from a function that this will replace to propagate them. This is strictly needed only for wave sensitivity, but will also copy over readnone and readonly. These don't currently see any use after codegen is done, but copying them over is cheap, matches expectations better, and causes no harm.

File: tools/clang/unittests/HLSL/ExtensionTest.cpp [hunk 8]
```diff
@@ -880,3 +888,43 @@ TEST_F(ExtensionTest, SamplerExtensionIntrinsic) {
...
+  );
+  std::string disassembly = c.Disassemble();
+
+  // Check that the wave op causes the break block to be retained
+  VERIFY_IS_TRUE(
+    disassembly.npos !=
+    disassembly.find("@dx.break.cond = internal constant [1 x i32] zeroinitializer"));
+  VERIFY_IS_TRUE(
+    disassembly.npos !=
+    disassembly.find("%1 = load i32, i32* getelementptr inbounds ([1 x i32], [1 x i32]* @dx.break.cond"));
```
[the remark this answers is not in the document]
Comment (pow2clk) [reply tools/clang/unittests/HLSL/ExtensionTest.cpp:919]: They both pass. I'd like to leave them in because it verifies the full path from load to conditional branch.

File: utils/hct/gen_intrin_main.txt [hunk 9]
```diff
@@ -245,38 +245,38 @@ bool [[rn]] CheckAccessFullyMapped(in uint_only status) : check_access_fully_map
...
+$type1 [[wv]] WaveMultiPrefixBitAnd(in any_int<> value, in uint<4> mask);
+$type1 [[wv]] WaveMultiPrefixBitOr(in any_int<> value, in uint<4> mask);
+$type1 [[wv]] WaveMultiPrefixBitXor(in any_int<> value, in uint<4> mask);
+uint [[wv]] WaveMultiPrefixCountBits(in bool value, in uint<4> mask);
+$type1 [[unsigned_op=WaveMultiPrefixUProduct,wv]] WaveMultiPrefixProduct(in numeric<> value, in uint<4> mask);
+$type1 [[unsigned_op=WaveMultiPrefixUSum,wv]] WaveMultiPrefixSum(in numeric<> value, in uint<4> mask);
+$type1 [[wv]] QuadReadLaneAt(in numeric<> value, in uint quadLane);
+$type1 [[wv]] QuadReadAcrossX(in numeric<> value);
+$type1 [[wv]] QuadReadAcrossY(in numeric<> value);
+$type1 [[wv]] QuadReadAcrossDiagonal(in numeric<> value);
```
[the remark this answers is not in the document]
Comment (pow2clk) [reply utils/hct/gen_intrin_main.txt:279]: I don't want to leave them in unnecessarily. I'll remove them.

Is exec mask the right criteria? I thought so.

File: lib/HLSL/HLOperations.cpp (hunk 6 above)
Comment (pow2clk) [reply lib/HLSL/HLOperations.cpp:505]: GetHLFullName ends with a "." The result of this is  "dx.hl.op.waveOp." 

Anything would be enough to differentiate these. I was patterning it after "binop"

File: lib/HLSL/HLMatrixLowerPass.cpp (hunk 2 above)
Comment (pow2clk) [reply lib/HLSL/HLMatrixLowerPass.cpp:854]: It could. It's called a lot of places which is why I backed away from that. Sometimes it's called to replace a previous function, in which case the attribs can be passed along. In other cases, it's a totally new function.

The only current use of these function attributes are for errors during codegen and the wave-sensitivity during the dxbreak cleanup. They might be useful for other things going forward though.

File: lib/HLSL/DxilPreparePasses.cpp [hunk 10]
```diff
@@ -1142,14 +1142,14 @@ class CleanupDxBreak : public FunctionPass {
...
-      for (User *U : IF.users()) {
-        CallInst *CI = dyn_cast<CallInst>(U);
-        if (CI && IsCallWaveSensitive(CI))
-          CollectSensitiveBlocks(LInfo, CI, BreakFunc, SensitiveBBs);
+      if (&IF != &F && IF.getNumUses() && IF.isDeclaration() &&
+          (hlsl::GetHLOpcodeGroupByName(&IF) == HLOpcodeGroup::HLIntrinsic ||
+           hlsl::GetHLOpcodeGroup(&IF) == HLOpcodeGroup::HLExtIntrinsic)) {
+        for (User *U : IF.users()) {
+          CallInst *CI = dyn_cast<CallInst>(U);
+          if (CI && IsHLWaveSensitive(CI->getCalledFunction()))
```
[the remark this answers is not in the document]
Comment (pow2clk) [reply lib/HLSL/DxilPreparePasses.cpp:1150]: yup. oops.

File: lib/HLSL/HLOperations.cpp (hunk 6 above)
Comment (tex3d) [reply lib/HLSL/HLOperations.cpp]: Ok, that sounds good.

File: lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp (hunk 4 above)
Comment (tex3d) [reply lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp:996]: Good catch!

File: lib/HLSL/DxilPreparePasses.cpp [hunk 11]
```diff
@@ -1142,13 +1142,12 @@ class CleanupDxBreak : public FunctionPass {
     // For each wave operation, collect the blocks sensitive to it
     SmallPtrSet<BasicBlock *, 16> SensitiveBBs;
     for (Function &IF : M->functions()) {
-      if (&IF != &F && IF.getNumUses() && IF.isDeclaration() &&
-          (hlsl::GetHLOpcodeGroupByName(&IF) == HLOpcodeGroup::HLIntrinsic ||
-           hlsl::GetHLOpcodeGroup(&IF) == HLOpcodeGroup::HLExtIntrinsic)) {
+      HLOpcodeGroup opgroup = hlsl::GetHLOpcodeGroup(&IF);
+      if (&IF != &F && IF.getNumUses() && IF.isDeclaration() && IsHLWaveSensitive(&IF) &&
```
Comment (tex3d) [inline lib/HLSL/DxilPreparePasses.cpp:1146]: Now the only problem I have is that getting the group is somewhat expensive, while some other tests are quite cheap that could early out, such as these: `&IF != &F && IF.getNumUses() && IF.isDeclaration()`.
Also '!IF.use_empty()` is cheaper than `IF.getNumUses()` (which has to iterate the uses to count them).
Not a blocker though.
